File: backend/tests/builders.py
"""Test data builders."""

from __future__ import annotations

import datetime as dt

from powerguard.domain.entities import Anomaly, AnomalyMethod, Telemetry

DEVICE_ID = "powerguard-01"
BOOT_ID = "7fa31c09"
FIRMWARE = "0.1.0"


def telemetry(
    *,
    seq: int = 1,
    received_at: dt.datetime,
    device_id: str = DEVICE_ID,
    boot_id: str = BOOT_ID,
    voltage_v: float = 7.84,
    current_a: float = 0.417,
    energy_wh: float = 0.284,
    sampled_at: dt.datetime | None = None,
) -> Telemetry:
    return Telemetry(
        device_id=device_id,
        boot_id=boot_id,
        seq=seq,
        received_at=received_at,
        sampled_at=sampled_at,
        voltage_v=voltage_v,
        current_a=current_a,
        power_w=voltage_v * current_a,
        energy_wh=energy_wh,
    )


def anomaly(
    *,
    telemetry_id: int,
    detected_at: dt.datetime,
    device_id: str = DEVICE_ID,
    method: AnomalyMethod = AnomalyMethod.RULE,
    reasons: tuple[str, ...] = ("overcurrent_rule",),
    score: float | None = 0.9,
) -> Anomaly:
    return Anomaly(
        telemetry_id=telemetry_id,
        device_id=device_id,
        detected_at=detected_at,
        method=method,
        reasons=reasons,
        score=score,
    )

File: frontend/tests/builders.ts
/** Deterministic fixtures. No clock, no randomness, no network. */

import type {
  AnomalyDto,
  DeviceDto,
  HealthDto,
  TelemetryDto,
} from "../src/api/contract";

export const DEVICE_ID = "powerguard-01";
export const BOOT_ID = "7fa31c09";

const BASE = Date.UTC(2026, 0, 1, 0, 0, 0);

export function stamp(seconds: number): string {
  return new Date(BASE + seconds * 1000).toISOString().replace(/\.(\d{3})Z$/, ".$1Z");
}

export function telemetry(overrides: Partial<TelemetryDto> = {}): TelemetryDto {
  const id = overrides.id ?? 1;
  return {
    id,
    device_id: DEVICE_ID,
    boot_id: BOOT_ID,
    seq: id,
    sampled_at: null,
    received_at: stamp(id * 2),
    voltage_v: 7.84,
    current_a: 0.417,
    power_w: 3.269,
    energy_wh: 0.284,
    sensor_status: "ok",
    anomaly: null,
    ...overrides,
  };
}

/** `count` rows, newest first, exactly as the backend returns a page. */
export function telemetryPage(count: number, startId = 1): TelemetryDto[] {
  const rows = Array.from({ length: count }, (_unused, index) =>
    telemetry({ id: startId + index }),
  );
  return rows.reverse();
}

export function device(overrides: Partial<DeviceDto> = {}): DeviceDto {
  return {
    id: DEVICE_ID,
    firmware_version: "0.1.0",
    status: "online",
    first_seen_at: stamp(0),
    last_seen_at: stamp(120),
    latest: telemetry({ id: 60 }),
    ...overrides,
  };
}

export function anomaly(overrides: Partial<AnomalyDto> = {}): AnomalyDto {
  return {
    id: 1,
    telemetry_id: 1,
    device_id: DEVICE_ID,
    detected_at: stamp(4),
    method: "rule",
    score: 0.91,
    model_version: null,
    reasons: ["overcurrent_rule"],
    voltage_v: 7.84,
    current_a: 3.2,
    power_w: 25.1,
    energy_wh: 0.3,
    ...overrides,
  };
}

export function health(overrides: Partial<HealthDto> = {}): HealthDto {
  return {
    status: "ok",
    database: "ready",
    mqtt: "connected",
    model: "unavailable",
    version: "0.1.0",
    ...overrides,
  };
}
